cloudimg/ms_azure.py

```diff
@@ -0,0 +1,516 @@
+from datetime import datetime, timedelta
+import logging
+import os
+try:
+    from time import monotonic
+except ImportError:
+    # Python < 3.3
+    from monotonic import monotonic
+
+import attr
+
+from azure.storage.blob import (
+    AccountSasPermissions,
+    BlobServiceClient,
+    BlobSasPermissions,
+    generate_account_sas,
+    generate_blob_sas,
+    ResourceTypes,
+)
+
+from cloudimg.common import BaseService, PublishingMetadata
+
+log = logging.getLogger(__name__)
+
+
+def _get_values_from_connection_string(connection_str):
+    """
+    Get all the values from the connection string.
+
+    Args:
+        connection_string (str): The connection string of the
+                                 storage account
+
+    Returns:
+        A dictionary with all the key-values in the connection string
+    """
+    # Extraction
+    res = {}
+    val_list = connection_str.split(";")
+    for val in val_list:
+        if "=" in val:
+            entry = val.split("=", 1)
+            res[entry[0]] = entry[1]
+
+    # Validation
+    mandatory_keys = ["AccountName", "AccountKey"]
+    err_prefix = "Invalid connection string:"
+    if not res:
+        raise ValueError("%s No keyword elements found." % err_prefix)
+
+    for mkey in mandatory_keys:
+        if mkey not in res.keys():
+            raise ValueError("%s Missing key %s" % (err_prefix, mkey))
+        elif not res.get(mkey, None):
+            raise ValueError("%s Missing value for %s" % (err_prefix, mkey))
+
+    return res
+
+
+def _get_tags_filter_expression(tags):
+    """
+    Create a filter expression from the given tags for search.
+
+    The filter expression is a string with format:
+    ```
+    "tag1"='value1' and "tag2"='value2' ...
+    ```
+
+    Args:
+        tags (dict): Dictionary of strings with the tags to filter
+
+    Returns:
+        String with the filter expression.
+    """
+
+    # Create the filter expression like:
+    # "tag1"='value1" and "tag2" = 'value2" ...
+    filter_expression = " and ".join([
+            "\"%s\"='%s'" % (k, v)
+            for k, v in tags.items()
+        ])
+
+    # strip the last ' and '
+    return filter_expression.rstrip(" and ")
```

Comment:
`rstrip` is not needed with `join()`

Reply:
Oh, that's right!

Thanks!

Reply:
Now the method has just one line as intended, tyvm!